File: illtyped.ml
let safeHead list = 
    match list with
    | [] -> None
    | (h::t) -> Some h

let safeTail list =
    match list with
    | [] -> []
    | (h::t) -> t

let compose f g x = f (g x)

let ifthenelse c a b = if c then a else b

let cons x xs = x :: xs

let lt : int -> int -> bool = (<)

let eq : bool -> bool -> bool = (=)

let welltypedProgram = (snd) (((lt) ((snd) ((List.map,(fst) ((0,ifthenelse)) )) ) (((+)) (((fun id -> id)) ((begin match ((@)) ([]) ([])  with | [] -> (begin match [] with | [] -> 1; | (z9::z10) -> z9 end); | (z3::z4) -> let y3 = ((fun id -> id)) in (0) end)) ) ((if ((if ((begin match [] with | [] -> true; | (z11::z12) -> false end)) then ((begin match [] with | [] -> false; | (z15::z16) -> false end)) else (((||)) (true) (false) ))) then ((begin match (begin match [] with | [] -> []; | (z25::z26) -> z25 end) with | [] -> (List.length) ([]) ; | (z23::z24) -> (List.length) (z24)  end)) else (((~-)) (((-)) (0) (0) ) ))) ) ,true)) 

let illtypedProgram  = (snd) (((lt) ((snd) ((List.map,(fst) ((List.length,ifthenelse)) )) ) (((+)) (((fun id -> id)) ((begin match ((@)) ([]) ([])  with | [] -> (begin match [] with | [] -> 1; | (z9::z10) -> z9 end); | (z3::z4) -> let y3 = ((fun id -> id)) in (0) end)) ) ((if ((if ((begin match [] with | [] -> true; | (z11::z12) -> false end)) then ((begin match [] with | [] -> false; | (z15::z16) -> false end)) else (((||)) (true) (false) ))) then ((begin match (begin match [] with | [] -> []; | (z25::z26) -> z25 end) with | [] -> (List.length) ([]) ; | (z23::z24) -> (List.length) (z24)  end)) else (((~-)) (((-)) (0) (0) ) ))) ) ,true)) 

File: illtyped.hs
import Distribution.Simple.Utils

main = putStrLn "executed main"

len :: [a] -> Int
len = foldr (\_ n -> n+1) 0

(<) :: Int -> Int -> Bool
(<) = (Prelude.<)

(==) :: Bool -> Bool -> Bool
(==) = (Prelude.==)

ifthenelse :: Bool -> a -> a -> a
ifthenelse c a b = if c then a else b

welltypedProgram = (snd) ((((Main.<)) ((snd) ((map,(fst) ((0,ifthenelse)) )) ) (((+)) ((id) ((case ((++)) ([]) ([])  of [] -> (case [] of [] -> 1; (z9:z10) -> z9;); (z3:z4) -> let y3 = (id) in (0);)) ) ((if ((if ((case [] of [] -> True; (z11:z12) -> False;)) then ((case [] of [] -> False; (z15:z16) -> False;)) else (((||)) (True) (False) ))) then ((case (case [] of [] -> []; (z25:z26) -> z25;) of [] -> (len) ([]) ; (z23:z24) -> (len) (z24) ;)) else ((negate) (((-)) (0) (0) ) ))) ) ,True)) 

illtypedProgram  = (snd) ((((Main.<)) ((snd) ((map,(fst) ((len,ifthenelse)) )) ) (((+)) ((id) ((case ((++)) ([]) ([])  of [] -> (case [] of [] -> 1; (z9:z10) -> z9;); (z3:z4) -> let y3 = (id) in (0);)) ) ((if ((if ((case [] of [] -> True; (z11:z12) -> False;)) then ((case [] of [] -> False; (z15:z16) -> False;)) else (((||)) (True) (False) ))) then ((case (case [] of [] -> []; (z25:z26) -> z25;) of [] -> (len) ([]) ; (z23:z24) -> (len) (z24) ;)) else ((negate) (((-)) (0) (0) ) ))) ) ,True)) 